src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java

```diff
@@ -166,16 +222,63 @@
      */
     @CheckForNull
     private transient /*effectively final*/ Map<String, ContributorMetadataAction> pullRequestContributorCache;
+    /**
+     * The cache of the clone links.
+     */
     @CheckForNull
     private transient List<BitbucketHref> cloneLinks = null;
 
-    private static final Logger LOGGER = Logger.getLogger(BitbucketSCMSource.class.getName());
-
+    /**
+     * Constructor.
+     *
+     * @param id         the id.
+     * @param repoOwner  the repository owner.
+     * @param repository the repository name.
+     */
     @DataBoundConstructor
-    public BitbucketSCMSource(String id, String repoOwner, String repository) {
+    public BitbucketSCMSource(@CheckForNull String id, @NonNull String repoOwner, @NonNull String repository) {
         super(id);
+        this.serverUrl = BitbucketCloudEndpoint.SERVER_URL;
         this.repoOwner = repoOwner;
         this.repository = repository;
+        this.traits = new ArrayList<>();
+        this.traits.add(new BranchDiscoveryTrait(true, true));
+        traits.add(new OriginPullRequestDiscoveryTrait(EnumSet.of(ChangeRequestCheckoutStrategy.MERGE)));
+        traits.add(new ForkPullRequestDiscoveryTrait(EnumSet.of(ChangeRequestCheckoutStrategy.MERGE),
+                new ForkPullRequestDiscoveryTrait.TrustTeamForks()));
+    }
+
+    /**
+     * Migrate legacy serialization formats.
+     *
+     * @return {@code this}
+     * @throws ObjectStreamException if things go wrong.
+     */
+    @SuppressWarnings("ConstantConditions")
+    @SuppressFBWarnings(value = "RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE",
+                        justification = "Only non-null after we set them here!")
+    private Object readResolve() throws ObjectStreamException {
+        if (serverUrl == null) {
+            serverUrl = BitbucketEndpointConfiguration.get().readResolveServerUrl(bitbucketServerUrl);
```

Comment:
`repositoryType` should not have been persisted ever. It is supposed to be a cache not part of the persisted state